generators/templates/validationtest.gotpl

```diff
@@ -0,0 +1,77 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by protoc-gen-go-googlecetypes. DO NOT EDIT.
+// versions:
+// 	protoc       				 {{ .ProtocVersion }}
+// 	protoc-gen-go 				 {{ .ProtocGenGoVersion }}
+// 	protoc-gen-go-googlecetypes  {{ .LibraryVersion }}
+// source: {{ .SourceFile }}
+
+package {{ .Pkg }}_test
+
+import (
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/googleapis/google-cloudevents-go/{{ .SrcPath }}"
+	"github.com/googleapis/google-cloudevents-go/internal/testhelper"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+{{ range .DataTypes }}
+// Validate the type can parse test data.
+// Goals:
+// - "Loose" parsing confirms the expected library experience
+// - "Strict" parsing confirms:
+//   - no deleted or renamed fields in protos covered in test data
+//   - test data does not carry unknown fields
+func TestParsing{{ . }}(t *testing.T) {
+	cases := testhelper.FindTestData(t, "{{ . }}", "{{ $.TestDataPath }}")
+
+	for name, file := range cases {
+		t.Run(name, func(t *testing.T) {
+			data, err := os.ReadFile(file)
+			if err != nil {
+				t.Fatal("os.ReadFile:", err)
+			}
+
+			if ext := filepath.Ext(file); ext != ".json" {
+				t.Fatalf("test support for %q data not implemented", ext)
+			}
+
+			t.Run("loose", func(t *testing.T) {
+				out := {{ $.Pkg }}.{{ . }}{}
+				pj := protojson.UnmarshalOptions{DiscardUnknown: true}
+				if err := pj.Unmarshal(data, &out); err != nil {
+					t.Fatalf("protojson.Unmarshal: could not parse %q\n----%s\n----", file, data)
+				}
+			})
+
+			t.Run("strict", func(t *testing.T) {
+				cleanData, err := testhelper.CleanAcceptedUnknowns(data, "{{ . }}", "{{ $.TypePrefix }}")
```

Comment:
The strict testing with "known unknowns" removed is a cross-check that protos and test data don't fall out of sync in unexpected ways, such as compatibility breaks in the protos. From that standpoint, we should always run a strict test.

Are you thinking instead of running cleanup on every test, we should move the conditional evaluating whether cleanup should happen into the test function itself? I wanted to keep it out of the code generation process so if future types slip in with a problem property we can directly modify the code instead of needing to move through a code generation lifecycle.

Discarded unknowns testing is included because that's the primary recommendation we planned for sample usage. The alternative is to bring this cleanup process to the runtime data path, which may not be bad as that might be needed for Pub/Sub field duplication.

Reply:
I'm not sure what the objective is here, as I see two different goals:

a) Prevent the 2 already-not-valid-jsonproto messages from getting worse
b) Show how to work around the already-not-valid-jsonproto messages, so we can use a stricter parser.

I do not think this is the right place for (b) - that's better done in a sample.
For (a) - It seems there are two different types of test. 
* *Most* objects are being parsed strictly
* the two special ones are "compatibility" tests, not strict parsing tests.

I think making a distinction with the test name, and only "cleaning" data for the known-bad data types is clearer. To keep the template cleaner, you could add a `dataNeedsCleaning` param, so changes could be minimal.

If this seems a bit confusing, i'm happy to jump on a quick call to explain my perspective.

Reply:
Split testing into three pieces, with skips for products that can't support strict parsing without modification.

As part of this, I moved the "Clean" function to another file and split it up by type, and moved the conditional logic on special casing products into the template.